snuba/datasets/dataset.py

```diff
@@ -28,58 +28,101 @@ def get_min_columns(self) -> Sequence[str]:
 
 class Dataset(object):
     """
-    A dataset represent one or multiple entities in the Snuba data model.
+    A dataset represent a data model we can run a Snuba Query on.
+    A data model provides an abstract schema (today it is a flat table,
+    soon it will be a graph of Entities).
+    The dataset (later the Entity) has access to multiple Storage objects,
+    each one represents a table/view on the DB we can query.
     The class is a facade to access the components used to write on the
     data model and to query the entities.
-    To query the data model, it provides a schema (for one table or for
-    multiple joined tables), query processing features and query exension
-    parsing features.
-    To write it CAN provide a TableWriter, which has a schema as well and
-    provides a way to stream input from different sources and write them to
-    Clickhouse.
+
+    When processing a query, there are three main steps:
+    - dataset query processing. A series of QueryProcessors are applied to the
+      query before deciding which Storage to use. These processors are defined
+      by the dataset
+    - the Storage to run the query onto is selected and the query is transformed
+      into a Storage Query. This is done by a StorageQueryPlanBuilder. This object
+      produces a plan that includes the Query contextualized on the storage/s, the
+      list of processors to apply and the strategy to run the query (in case of
+      any strategy more complex than a single DB query like a split).
+    - storage query processing. A second series of QueryProcessors are applied
+      to the query. These are defined by the storage.
+
+    The architecture of the Dataset is divided in two layers. The highest layer
+    provides the logic we use to deal with the data model. (writers, query processors,
+    query planners, etc.). The lowest layer incldues simple objects that defines
+    the query itself (Query, Schema, RelationalSource). The lop layer object access and
+    manipulate the lower layer objects.
     """
 
     def __init__(
         self,
-        dataset_schemas: DatasetSchemas,
         *,
-        table_writer: Optional[TableWriter] = None,
+        storages: Sequence[Storage],
```

Comment:
My rough idea to introduce entities is that they will be an intermediate level between the dataset and the Storage. So this will be a sequence of entities and writable_storage will be on the Entity. QueryPlanBuilder responsibility would likely be split between dataset and Entity. Here it will seem trivial for events and discover but we need to take joins into account where it will be more complex.